common/flatpak-run.c

```diff
@@ -3323,60 +3324,87 @@ flatpak_run_app (const char     *app_ref,
   if (app_deploy != NULL)
     {
       g_autofree const char **previous_ids = NULL;
-      gsize len = 0, i;
+      gsize len = 0;
+      int i;
+
+      real_app_id_dir = flatpak_get_data_dir (app_ref_parts[1]);
 
+      previous_app_id_dirs = g_ptr_array_new_with_free_func (g_object_unref);
       app_files = flatpak_deploy_get_files (app_deploy);
 
-      g_autoptr (GFile) current_data_dir = flatpak_get_data_dir (app_ref_parts[1]);
       previous_ids = flatpak_deploy_data_get_previous_ids (app_deploy_data, &len);
-      i = len;
-      while (i > 0)
+      for (i = len - 1; i >= 0; i--)
         {
-          g_autoptr (GFile) previous_data_dir = NULL;
-          g_autoptr (GError) local_error = NULL;
-          g_autofree char *previous_data_path = NULL;
+          g_autoptr(GFile) previous_app_id_dir = NULL;
+          g_autoptr(GFileInfo) previous_app_id_dir_info = NULL;
+          g_autoptr(GError) local_error = NULL;
+          gboolean create_symlink;
+
+          previous_app_id_dir = flatpak_get_data_dir (previous_ids[i]);
+          previous_app_id_dir_info = g_file_query_info (previous_app_id_dir,
+                                                        G_FILE_ATTRIBUTE_STANDARD_IS_SYMLINK ","
+                                                        G_FILE_ATTRIBUTE_STANDARD_SYMLINK_TARGET,
+                                                        G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS,
+                                                        cancellable,
+                                                        &local_error);
+          if (previous_app_id_dir_info == NULL && !g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+            {
+              g_propagate_error (error, g_steal_pointer (&local_error));
+              return FALSE;
+            }
 
-          if (g_file_query_exists (current_data_dir, cancellable))
-            break;
+          g_clear_error (&local_error);
 
-          previous_data_dir = flatpak_get_data_dir (previous_ids[--i]);
-          if (!g_file_query_exists (previous_data_dir, cancellable))
-            continue;
+          /* If the previous data dir exists, but the current one doesn't, rename the old
+             one to the new one. (We need to be careful to not move some old symlink over
+             onto the current directory.) */
+          if (previous_app_id_dir_info != NULL && !g_file_info_get_is_symlink (previous_app_id_dir_info) &&
```

Comment:
Hmm, I'm not sure this is great. I can imaging people replacing e.g. ~/.var/app/com.valvesoftware.Steam with a symlink to some other location. If so we should not ignore it but move it to the new name.

Reply:
I think you can just drop the symlink check. In case com.valvesoftware.Steam is a symlink to /mnt/data/steam, then we should just move that symlink to the new name, so the rename is correct.